### ISSUE TITLE
Duplicating Blocks

### ISSUE THREAD

**Issue by peterbedor** [spicywebau/craft-neo#138]
It seems that in some cases, and I'm still trying to figure out exactly when, Neo blocks appear duplicated in the CMS as well as front end.  I did a little digging and found that it was, in my case,  caused by the structure elements join statement.

For anyone else, I have a temporary fix that seems to work so far:

Adding `$this->query->groupBy('{{elements}}.uid');` to the `beforePrepare` method in `BlockQuery.php`

Still testing it out, but figured this might be helpful.

**Comment by ttempleton** [comment]
Hi @peterbedor,

I've as yet been unable to reproduce this, sorry.  Were you able to find out any more about the cases where this issue appears?

**Comment by peterbedor** [comment]
@ttempleton Yeah, it's really weird and I've also been unable to figure out exactly when/how/why this happens.  Someone in another thread mentioned that it happens to them when [they use drafts](https://github.com/spicywebau/craft-neo/issues/127#issuecomment-425572403).  I'm using drafts all over and it's only happened in two channels.

I really wish I had more info.

**Comment by ttempleton** [comment]
Is your Craft installation multi-site and are those two channels using Neo fields set to manage blocks on a per-site basis?  I've been able to reproduce the issue in that case (though I'm only experiencing it on blocks associated with the first site).  Four of each block, just as you described in [#127](https://github.com/spicywebau/craft-neo/issues/127#issuecomment-420812406), and I can confirm that your fix has removed the duplication for me.

**Comment by peterbedor** [comment]
Not using multi-site or locales.  It's a pretty basic set up.  I can remove the fix locally and see what happens and try to get you some more information.

**Comment by ttempleton** [comment]
I'm curious to know what could have caused the issue for you, then, but the most important thing is that your fix does resolve the issue for you and it does resolve the duplication I was seeing, so I've now added it to the project code.  I'll close this issue now; it probably isn't necessary to look further into it unless you really want to, but if you do look further into it and find out any more, by all means let us know.  Thanks for reporting and fixing this!

**Comment by peterbedor** [comment]
Hey @ttempleton - I just pulled in the most recent version of Neo and noticed that my duplication issues are back.  I saw that you removed the line in [this commit](https://github.com/spicywebau/craft-neo/commit/eb7c20ad8acd33d2d6ec061ceb746062c8260fa7).  Was this intentional?  Was it causing other issues?

**Comment by ttempleton** [comment]
There were issues with it in certain cases, yeah; but I also found that the core problem was multiple block structures sometimes being created for a given field/owner/site combination and therefore blocks being associated with multiple structures in those cases.  I ended up with about twenty of each block appearing for one particular entry; it was quite nightmarish 😄.  In my case, it was caused by a failure to apply an owner site ID to an existing block structure when enabling multiple sites and block management on a per-site basis.  I'm unsure of what specifically would've caused it for you, since you don't have multiple sites, but given your initial report that the duplication was occurring with the structure elements join, I do believe that, somehow, you've ended up with multiple block structures for the affected entries.

Now, the other changes in that commit have prevented the issue from reoccurring for me (and hopefully for you, too) -- but only for any entries/categories/etc. that are not already affected.  The (potentially) good news is that, when saving an affected entry as a new entry, the new entry's Neo fields will have new block structures created and therefore should function normally.  I'm not sure how much of an inconvenience it'd be to resolve your situation that way, though, as I'm not sure how many affected entries you have, what entry fields you might use that link to affected entries, and such.

I think it'd be possible to have some form of block structure cleanup in the plugin; I unfortunately won't be able to investigate this until next week, as I'm on leave this week, but can look into it once I'm back in the office.

(Also, if you do try 'save as new entry' with the latest Neo version and still see duplicate blocks in the new entry, then let us know and I'll reopen this issue.)
